runtime/vm/resolvesupport.cpp

```diff
@@ -1847,6 +1847,106 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+			/* Another thread is currently trying to resolve the entry */
+			omrthread_monitor_enter(vm->constantDynamicMutex);
+			omrthread_monitor_wait(vm->constantDynamicMutex);
+			omrthread_monitor_exit(vm->constantDynamicMutex);
+			goto retry;
+		}
+	}
+
+	/* Enter if not previously resolved */
+	if (!resolved) {
+		J9Class *ramClass = ramCP->ramClass;
+		J9ROMClass *romClass = ramClass->romClass;
+		J9ROMConstantDynamicRef *romConstantRef = (J9ROMConstantDynamicRef*)(J9_ROM_CP_FROM_CP(ramCP) + cpIndex);
+		J9SRP *callSiteData = (J9SRP *) J9ROMCLASS_CALLSITEDATA(romClass);
+		U_16 *bsmIndices = (U_16 *) (callSiteData + romClass->callSiteCount);
+		U_16 *bsmData = bsmIndices + romClass->callSiteCount;
+		U_32 bsmIndex = romConstantRef->bsmIndexAndCpType >> J9DescriptionCpTypeShift;
+		J9ROMNameAndSignature* nameAndSig = SRP_PTR_GET(&romConstantRef->nameAndSignature, J9ROMNameAndSignature*);
+		J9MemoryManagerFunctions *gcFuncs = vm->memoryManagerFunctions;
+
+		/* Set exception to current threadObject as a mutex to prevent other threads from entering */
+		if ((ramCPEntry->exception != vmThread->threadObject) &&
+			(0 == gcFuncs->j9gc_objaccess_staticCompareAndSwapObject(vmThread, ramClass, &ramCPEntry->exception, NULL, vmThread->threadObject))
+		) {
+			goto retry;
+		}
+
+		/* Walk bsmData */
+		for (U_32 i = 0; i < bsmIndex; i++) {
+			bsmData += bsmData[1] + 2;
+		}
+
+		/* Invoke BootStrap method*/
+		sendResolveConstantDynamic(vmThread, ramCP, cpIndex, nameAndSig, bsmData);
+		value = (j9object_t)vmThread->returnValue;
+
+		/* Check if entry resolved by nested constantDynamic calls */
+		if (ramCPEntry->exception != vmThread->threadObject) {
+			goto retry;
+		}
+
+		j9object_t exceptionObject = NULL;
+		if (NULL != vmThread->currentException) {
+			/* Allocate the exception object in tenure */
+			exceptionObject = gcFuncs->j9gc_objaccess_asConstantPoolObject(
+											vmThread,
+											vmThread->currentException,
+											J9_GC_ALLOCATE_OBJECT_TENURED | J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+			if (NULL == exceptionObject) {
+				setHeapOutOfMemoryError(vmThread);
+				exceptionObject = vmThread->outOfMemoryError;
+			}
+		} else if (NULL == value) {
+			/* Java.lang.void is used as a special flag to indicate null reference */
+			exceptionObject = vm->voidReflectClass->classObject;
+		} else {
+			value = gcFuncs->j9gc_objaccess_asConstantPoolObject(
+									vmThread,
+									value,
+									J9_GC_ALLOCATE_OBJECT_TENURED | J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE | J9_GC_ALLOCATE_OBJECT_HASHED);
+
+			if (NULL == value) {
+				setHeapOutOfMemoryError(vmThread);
+				exceptionObject = vmThread->outOfMemoryError;
+			} else {
+				ramCPEntry->value = value;
+			}
+		}
+
+		ramCPEntry->exception = exceptionObject;
+		omrthread_monitor_enter(vm->constantDynamicMutex);
+		omrthread_monitor_notify_all(vm->constantDynamicMutex);
+		omrthread_monitor_exit(vm->constantDynamicMutex);
```

Comment:
This is not a legal use of wait/notify.  Let's discuss tomorrow.

Reply:
It would be better to stash the two written values in locals, and update the ramCPEntry while holding the monitor.

Reply:
I know it is a better practice to update the ramCPEntry within the monitor, but not sure how much overhead this would add in a multi-thread case when parallel resolve takes place

Reply:
None at all - it's the same code (modulo a few local variable accesses), just shuffled around.

Reply:
I've made the change to update the cp entry values inside the monitor scope, we can change this in the future if this becomes a performance issue